internal/app/frontend/frontend_service.go

```diff
@@ -354,7 +354,62 @@ func doWatchAssignments(ctx context.Context, id string, sender func(*pb.Assignme
 // AcknowledgeBackfill is used to notify OpenMatch about GameServer connection info.
 // This triggers an assignment process.
 func (s *frontendService) AcknowledgeBackfill(ctx context.Context, req *pb.AcknowledgeBackfillRequest) (*pb.Backfill, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+	if req.GetBackfillId() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, ".BackfillId is required")
+	}
+	if req.GetAssignment() == nil {
+		return nil, status.Errorf(codes.InvalidArgument, ".Assignment is required")
+	}
+
+	m := s.store.NewMutex(req.GetBackfillId())
```

Comment:
Note that on line 406 we execute `UpdateBackfill` and Backfill internal protobuf format includes also ticketed associated with this Backfill. So this lock is necessary here.
https://github.com/googleforgames/open-match/pull/1293/files#diff-fbf6a8aad66158048250071262ad0c5fd8c08c9034840c5802784888408881f3R406

Reply:
I do agree that the lock is necessary here  prevent the backfill from being modified from concurrent calls. I am not sure if it is sufficient. Can you explain how it protects the UpdateAssignments call from modifying the tickets while they are being updated from the AcknowledgeBackfill call?

Reply:
Tickets are immutable, so we don't need to keep track of their state. They can only be assigned, but searchFields could not be updated, this is the main difference with Backfill which has generation (version) number.

Reply:
I did not mention updating the searchFields at all. 

TicketA is a part of BackfillA. You receive a call to AcknowledgeBackfill for BackfillA - and are updating the assignment on the TicketA by making UpdateAssignments call - Line 389.

Concurrently, the Backend receives AssignTickets method setting Assignment for TicketA to a different assignment. 

Is there anything that prevents the AssignTickets method from calling  UpdateAssignments call on TicketA concurrently with AcknowledgeBackfill calliong UpdateAssignments for TicketA?

Reply:
Actually this is not possible to have one ticket being associated to two backfills or used in a normal match. Evaluator would not make this happen. Only one match would be selected with a ticketA and later on it would be moved into `pendingRelease` tickets not visible by `QueryTickets` later on. We are working on scale tests it will let us know if concurrent issues like this would arise.